serialization.rst

```diff
@@ -0,0 +1,308 @@
+  property [1]_, but it should **not** be set as property in the returned
+  object.
+
+- ``any other string`` — defines the class name that the BSON array or BSON
+  object should be deserialized at.
+
+  If the class implements the ``MongoDB\BSON\Unserializable`` interface,
+  either directly, or indirectly via ``MongoDB\BSON\Persistable``, then
+  the properties of the BSON document, **including** the ``__pclass`` property
+  if it exists, are sent as an associative array to the ``bsonUnserialize``
+  function to initialise the object's properties.
+
+  If the class does not implement the ``MongoDB\BSON\Unserializable``
+  interface, then an ``MongoDB\Driver\Exception\UnexpectedValueException``
+  exception is thrown.
+
+  If the named class is different from the ``__pclass`` key's value, then the
+  ``__pclass`` value is ignored and the class name from the type map is used.
+  The properties of the BSON document are sent to ``bsonUnserialize`` as per
+  above.
+
+TypeMaps
+--------
+
+TypeMaps can be set through the ``setTypeMap()`` on a
+``MongoDB\Driver\Cursor`` object, or the ``$typeMap`` argument of
+``MongoDB\BSON\toPHP()`` (previously, ``MongoDB\BSON\toArray()``). Each of the
+three classes (``root``, ``document`` and ``array``) can be individually set.
+
+If the value in the map is ``NULL``, it means the same as the *default* value
+for that item.
+
+Examples
+--------
+
+In these examples, ``MyClass`` does **not** implement any interface,
+``YourClass`` implements ``MongoDB\BSON\Unserializable`` and ``OurClass``
+implements ``MongoDB\BSON\Persistable``.
+
+The ``bsonUnserialize()`` method of ``YourClass`` and ``OurClass``
+iterate over the array and set the properties without modifications. It
+**also** sets the ``unserialized = true`` property::
+
+    function bsonUnserialize( array $map )
+    {
+        foreach ( $map as $k => $value )
+        {
+            $this->$k = $value;
+        }
+        $this->unserialized = true;
+    }
+
+::
+
+    /* typemap: [] (all defaults) */
+    { foo: 'yes', 'bar' : false }
+      -> stdClass { $foo => 'yes', $bar => false }
+
+    { foo: 'no', 'array' : [ 5, 6 ] }
+      -> stdClass { $foo => 'no', $array => [ 5, 6 ] }
+
+    { foo: 'no', 'obj' : { 'embedded' => 3.14 } }
+      -> stdClass { $foo => 'no', $obj => stdClass { $embedded => 3.14 } }
+
+    { foo: 'yes', '__pclass': 'MyClass' }
+      -> stdClass { $foo => 'yes', $__pclass => 'MyClass' }
+
+    { foo: 'yes', '__pclass': Binary(0x80, 'MyClass') }
+      -> MongoDB\Driver\Exception\UnexpectedValueException("class does not implement unserializable interface")
+
+    { foo: 'yes', '__pclass': Binary(0x80, 'YourClass') }
+      -> MyClass { $foo => 'yes', $__pclass => Binary(0x80, 'YourClass'), $unserialized => true }
+
+    { foo: 'yes', '__pclass': Binary(0x44, 'YourClass') }
+      -> stdClass { $foo => 'yes', $__pclass => Binary(0x44, 'YourClass') }
+
+::
+
+    /* typemap: [ 'document' => 'MyClass' ] */
```

Comment:
Does "document" have any significance here? There are no embedded documents in the example BSON.